File: app/lib/algorithms/simulated_annealing.rb
# frozen_string_literal: true

module SimulatedAnnealing
  module_function

  def simulated_annealing(arr, max_iterations = 1000, initial_temp = 1000, cooling_rate = 0.995)
    current = arr.shuffle
    best = current.dup
    temp = initial_temp

    max_iterations.times do
      neighbor = current.dup
      i, j = rand(current.length), rand(current.length)
      neighbor[i], neighbor[j] = neighbor[j], neighbor[i]

      current_cost = current.each_with_index.sum { |x, i| x * i }
      neighbor_cost = neighbor.each_with_index.sum { |x, i| x * i }

      delta = neighbor_cost - current_cost
      if delta < 0 || Math.exp(-delta / temp) > rand
        current = neighbor
        current_cost = neighbor_cost if neighbor_cost < current_cost
      end

      best = current.dup if current_cost < best.each_with_index.sum { |x, i| x * i }
      temp *= cooling_rate
    end
    best
  end
end

File: app/js/lib/algorithms/simulated_annealing.js
const SimulatedAnnealing = {
  simulated_annealing(arr, maxIterations = 1000, initialTemp = 1000, coolingRate = 0.995) {
    let current = [...arr].sort(() => Math.random() - 0.5);
    let best = [...current];
    let temp = initialTemp;

    const cost = (arr) => arr.reduce((sum, x, i) => sum + x * i, 0);
    let currentCost = cost(current);

    for (let i = 0; i < maxIterations; i++) {
      const neighbor = [...current];
      const idx1 = Math.floor(Math.random() * neighbor.length);
      const idx2 = Math.floor(Math.random() * neighbor.length);
      [neighbor[idx1], neighbor[idx2]] = [neighbor[idx2], neighbor[idx1]];

      const neighborCost = cost(neighbor);
      const delta = neighborCost - currentCost;

      if (delta < 0 || Math.exp(-delta / temp) > Math.random()) {
        current = neighbor;
        currentCost = neighborCost;
      }

      if (currentCost < cost(best)) best = [...current];
      temp *= coolingRate;
    }
    return best;
  }
};

module.exports = SimulatedAnnealing;
